xmtp_mls/src/client.rs

```diff
@@ -68,6 +81,92 @@ where
         }
     }
 
+    // TODO: Remove this and figure out the correct lifetimes to allow long lived provider
+    fn mls_provider(&self) -> XmtpOpenMlsProvider {
+        XmtpOpenMlsProvider::new(&self.store)
+    }
+
+    pub async fn register_identity(&self) -> Result<(), ClientError> {
+        // TODO: Mark key package as last_resort in creation
+        let last_resort_kp = self.identity.new_key_package(&self.mls_provider())?;
+        let last_resort_kp_bytes = last_resort_kp.tls_serialize_detached()?;
+
+        self.api_client
+            .register_installation(last_resort_kp_bytes)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn top_up_key_packages(&self) -> Result<(), ClientError> {
+        let key_packages: Result<Vec<Vec<u8>>, ClientError> = (0..KEY_PACKAGE_TOP_UP_AMOUNT)
+            .map(|_| -> Result<Vec<u8>, ClientError> {
+                let kp = self.identity.new_key_package(&self.mls_provider())?;
+                let kp_bytes = kp.tls_serialize_detached()?;
+
+                Ok(kp_bytes)
+            })
+            .collect();
+
+        self.api_client.upload_key_packages(key_packages?).await?;
+
+        Ok(())
+    }
+
+    async fn get_all_active_installation_ids(
+        &self,
+        wallet_addresses: Vec<String>,
+    ) -> Result<Vec<Vec<u8>>, ClientError> {
+        let update_mapping = self
+            .api_client
+            .get_identity_updates(0, wallet_addresses)
+            .await?;
+
+        let mut installation_ids: Vec<Vec<u8>> = vec![];
+
+        for (_, updates) in update_mapping {
+            let mut tmp: HashSet<Vec<u8>> = HashSet::new();
+            for update in updates {
+                match update {
+                    IdentityUpdate::Invalid => {}
+                    IdentityUpdate::NewInstallation(new_installation) => {
+                        // TODO: Validate credential
+                        tmp.insert(new_installation.installation_id);
+                    }
+                    IdentityUpdate::RevokeInstallation(revoke_installation) => {
+                        tmp.remove(&revoke_installation.installation_id);
+                    }
```

Comment:
@neekolas Something to keep in mind for when we implement revocation, but it might make sense to go through all revocations sequentially after we go through all `NewInstallations`, so that revocations can't be undone depending on the ordering.

Reply:
FWIW, these are all correctly ordered coming from the server. And there is no way to re-register an installation.

Reply:
I just realized that this is getting incremental updates since the last fetch, so we can't stop reordering issues regardless of how we process them here.

There is no timestamp inside the wallet signature text currently, so it's possible for a malicious client to reupload an existing registration. I guess if we make sure that the server disallows this when we implement it, in addition to maintaining consistent ordering, it should be fine. But it would be preferable to have Defense in Depth where possible, such as persisting that an installation was revoked in the DB. But these are all minor points for later I guess

Reply:
If a malicious client tries to re-upload an existing registration, the node won't do anything. We have the installation_id as the primary key in the DB, so you can only register one installation once.